homeassistant/components/avreceiver/__init__.py

```diff
@@ -0,0 +1,172 @@
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the AV Receiver component."""
+    if DOMAIN not in config:
+        return True
+    host = config[DOMAIN][CONF_HOST]
+    entries = hass.config_entries.async_entries(DOMAIN)
+    if not entries:
+        # Create new entry based on config
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": "import"}, data={CONF_HOST: host}
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up AV Receiver from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    host = entry.data[CONF_HOST]
+    controller = None
+    try:
+        controller = await factory(host)
+        await controller.init()
+        entry.unique_id = (
+            f"{DOMAIN}-{controller.serial_number or controller.mac or controller.host}"
+        )
+    except Exception as error:
+        if controller:
+            await controller.disconnect()
+        raise ConfigEntryNotReady from error
+
+    # Disconnect when shutting down
+    async def disconnect_avreceiver(event):
+        await controller.disconnect()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, disconnect_avreceiver)
+
+    controller_manager = AVRManager(hass, controller)
+    await controller_manager.connect_listeners()
+
+    unsub = entry.add_update_listener(update_listener)
+
+    zones = [
+        CONF_ZONE1,
+        CONF_ZONE2 if entry.options.get(CONF_ZONE2) else "",
+        CONF_ZONE2 if entry.options.get(CONF_ZONE3) else "",
+        CONF_ZONE2 if entry.options.get(CONF_ZONE4) else "",
+    ]
+    zones = {
+        name: getattr(controller, name)
+        for name in zones
+        if getattr(controller, name, None)
+    }
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        CONTROLLER: controller_manager,
+        MEDIA_PLAYER_DOMAIN: zones,
+        UNSUB_UPDATE_LISTENER: unsub,
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    await hass.data[DOMAIN][entry.entry_id][CONTROLLER].disconnect()
+
+    hass.data[DOMAIN][entry.entry_id][UNSUB_UPDATE_LISTENER]()
+
+    entity_registry = await er.async_get_registry(hass)
+    entries = er.async_entries_for_config_entry(entity_registry, entry.entry_id)
+    zone_names = [CONF_ZONE1, CONF_ZONE2, CONF_ZONE3, CONF_ZONE4]
+    for zone in entries:
+        for name in zone_names:
+            if zone.unique_id == f"{entry.unique_id}-{name}":
+                entity_registry.async_remove(zone.entity_id)
+
+    hass.data[DOMAIN].pop(entry.entry_id)
+
+    return await hass.config_entries.async_forward_entry_unload(
+        entry, MEDIA_PLAYER_DOMAIN
+    )
```

Comment:
The twice is because you remove the "main" zone on line 115 with `entity_registry.async_remove(zone.entity_id)`
and again when you call `return await hass.config_entries.async_forward_entry_unload` on line 119.

Just remove `CONF_ZONE1` from the list on line 111.

Reply:
Besides I would first do the `await hass.config_entries.async_forward_entry_unload`, and then remove the left over aditional zones.
So move lines 119-121 to 108

Reply:
If it then still leaves "ghost" entities, there is probably something wrong with the unique_id matching.
easiest way is to add a _LOGGER.error() with the zone.unique_id on line 113 and then just look in the log how the unique id's look like and if they match with your f"{entry.unique_id}-{name}"

Reply:
Very nice!  Errors are gone but I still have ghost entities and will investigate further.  
![image](https://user-images.githubusercontent.com/34154542/104386490-2bde4700-54ea-11eb-9e45-25114079057c.png)

Reply:
Now `async_foward_entry_unload` is above the entity_registry stuff.  Another hint is that the entity_registry still contains both the main and zone2 entities.  I think I must have screwed up reference to the entity somewhere.